api/src/main/java/me/lokka30/treasury/api/common/event/Completion.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * This file is/was part of Treasury. To read more information about Treasury such as its licensing, see <https://github.com/lokka30/Treasury>.
+ */
+
+package me.lokka30.treasury.api.common.event;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.Consumer;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class Completion {
+
+    public static Completion completed() {
+        return new Completion(true);
+    }
+
+    public static Completion completedExceptionally(@NotNull Throwable error) {
+        return new Completion(error);
+    }
+
+    public static Completion completedExceptionally(@NotNull List<@NotNull Throwable> errors) {
+        return new Completion(errors);
+    }
+
+    public static Completion join(Completion... other) {
+        List<Throwable> errors = new ArrayList<>();
+        for (Completion completion : other) {
+            completion.waitCompletion();
+            if (completion.getErrors() != null) {
+                errors.addAll(completion.getErrors());
+            }
+        }
+        return errors.isEmpty()
+                ? Completion.completed()
+                : Completion.completedExceptionally(errors);
+    }
+
+    private CountDownLatch latch = new CountDownLatch(1);
+    private List<Throwable> errors;
+    private Consumer<@Nullable List<@NotNull Throwable>> completedTask;
+
+    public Completion() {
+
+    }
+
+    private Completion(boolean completed) {
+        if (completed) {
+            this.latch.countDown();
+        }
+    }
+
+    private Completion(@NotNull List<@NotNull Throwable> errors) {
+        this.errors = Objects.requireNonNull(errors, "errors");
+        this.latch.countDown();
+    }
+
+    private Completion(@NotNull Throwable error) {
+        this(Collections.singletonList(Objects.requireNonNull(error, "error")));
+    }
+
+    public void complete() {
+        if (latch.getCount() == 0) {
+            throw new IllegalArgumentException("Completion already completed");
```

Comment:
Maybe a [IllegalStateException](https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalStateException.html) is better for this use
> Signals that a method has been invoked at an illegal or inappropriate time. In other words, the Java environment or Java application is not in an appropriate state for the requested operation.